### ISSUE TITLE
fails to delete container with run --rm

### ISSUE THREAD

**Issue by hickeng** [vmware/vic#1645]
There is a timing hole when deleting containers via `run --rm`. The rm occurs as soon as the connection drops in this scenario. Because the shutdown is initiated from within the container and there's an actual OS shutdown, the VM may not have finished powering off by the time the rm is dispatched.

Recreate
```
docker run -it --rm busybox /bin/bash
# exit
exit
Error response from daemon: server error from portlayer : [DELETE /containers/{id}][500] containerRemoveInternalServerError
```
port layer
```
time="2016-07-17T22:18:49Z" level=info msg="Finished copying stdout stream for container 97633e2f38093f0a3d069d822d67c034c2079dd216aeca305b5ba2e53a19b215" 
time="2016-07-17T22:18:49Z" level=info msg="Finished copying stderr stream for container 97633e2f38093f0a3d069d822d67c034c2079dd216aeca305b5ba2e53a19b215" 
time="2016-07-17T22:18:49Z" level=info msg="Done copying stdin" 
time="2016-07-17T22:18:49Z" level=debug msg="[BEGIN] [github.com/vmware/vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).RemoveContainerHandler] Containers.RemoveContainerHandler" 
time="2016-07-17T22:18:49Z" level=debug msg="[BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*Container).Remove] Container.Remove" 
time="2016-07-17T22:18:49Z" level=error msg="Operation failed: The attempted operation cannot be performed in the current state (Powered on). - (&{{} %!!(MISSING)s(*types.InvalidPowerState=&{{{{<nil> []}}} poweredOff poweredOn}) The attempted operation cannot be performed in the current state (Powered on).})" 
time="2016-07-17T22:18:49Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/portlayer/exec.(*Container).Remove] [418.087372ms] Container.Remove" 
time="2016-07-17T22:18:49Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).RemoveContainerHandler] [419.482633ms] Containers.RemoveContainerHandler" 
```

This may need some logic to allow for a delayed retry in this scenario - we should run some loop tests to determine how frequently this is seen.

**Comment by mdubya66** [comment]
assigning to jzt for initial evaluation

**Comment by jzt** [comment]
@hickeng This relies on `ContainerWait` (#384) being implemented. `docker run --rm` uses it to wait for the status code from a container exiting.

**Comment by jzt** [comment]
@cgtexmex has a [PR](https://github.com/vmware/vic/pull/2304) out for `ContainerWait`. I will revisit this once that is merged.

**Comment by jzt** [comment]
I will set it to in progress for now so I don't forget.

**Comment by jzt** [comment]
This use case appears to be fixed by @cgtexmex's `ContainerWait` PR:

```
vagrant@devbox $ docker -H 192.168.218.235:2376 --tls run --rm -it busybox /bin/sh
/ # exit
```

We can see in the logs that `ContainerWait` is called by the client once `Attach` completed, before its call to remove:

```
time="2016-05-26T14:00:19Z" level=info msg="container attach: stderr (ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73) done: %!s(MISSING)"
time="2016-05-26T14:00:19Z" level=error msg="container attach: stdin (ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73): context canceled"
time="2016-05-26T14:00:19Z" level=debug msg="* stdout attach end Thu May 26 14:00:19 UTC 2016"
time="2016-05-26T14:00:19Z" level=info msg="container attach: stdout (ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73) done: %!s(MISSING)"
time="2016-05-26T14:00:19Z" level=info msg="container attach:  cleaned up connections to ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73."
time="2016-05-26T14:00:19Z" level=error msg="container attach error: context canceled"
time="2016-05-26T14:00:19Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerAttach] [22.707685041s] ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73"
time="2016-05-26T14:00:19Z" level=error msg="Handler for POST /v1.23/containers/ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73/attach returned error: context canceled"
2016/05/26 14:00:19 http: response.WriteHeader on hijacked connection
2016/05/26 14:00:19 http: response.Write on hijacked connection
time="2016-05-26T14:00:19Z" level=error msg="container attach: stdin err: read tcp 192.168.218.235:2376->192.168.218.180:54120: use of closed network connection"
time="2016-05-26T14:00:19Z" level=debug msg="Calling POST /v1.23/containers/ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73/wait"
time="2016-05-26T14:00:19Z" level=debug msg="Client and server don't have the same version (client: 1.11.2, server: 1.22)"
time="2016-05-26T14:00:19Z" level=debug msg="[BEGIN] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerWait] name(ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73):timeout(-1s)"
time="2016-05-26T14:00:19Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerWait] [275.257247ms] name(ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73):timeout(-1s)"
time="2016-05-26T14:00:19Z" level=debug msg="Calling GET /v1.23/containers/ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73/json"
time="2016-05-26T14:00:19Z" level=debug msg="Client and server don't have the same version (client: 1.11.2, server: 1.22)"
time="2016-05-26T14:00:19Z" level=debug msg="[BEGIN] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerInspect] ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73"

[ snip container inspect logs for brevity ]

time="2016-05-26T14:00:19Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerInspect] [985.19µs] ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73"
time="2016-05-26T14:00:19Z" level=debug msg="Calling DELETE /v1.23/containers/ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73?v=1"
time="2016-05-26T14:00:19Z" level=debug msg="Client and server don't have the same version (client: 1.11.2, server: 1.22)"
time="2016-05-26T14:00:19Z" level=debug msg="[BEGIN] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerRm] ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73"
time="2016-05-26T14:00:19Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Container).ContainerRm] [110.808738ms] ebdf4cfb94458b5f88b8ec50133e773bc197b640ca0ac43c37823bedc658fd73"
```

There is still an issue with short-lived containers along these lines. I've created a new issue to track it here:  https://github.com/vmware/vic/issues/2344

Closing this as done.
